### PR TITLE
Path-width decomposition for our graph

### PR DESCRIPTION
**Pull request by elproblema**

### CODE REVIEW

File: src/composed-graph.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        if (_matchings[v] < n) throw CorrectnessException(v, _matchings[v], u, this);
+        if (_matchings[u] < n) throw CorrectnessException(u, _matchings[u], v, this);
+        _matchings[u] = v;
+        _matchings[v] = u;
+    }
+}
+
+ComposedGraph::CorrectnessException::CorrectnessException(vertex_t outer_vertex, ComposedGraph *G):
+_G(G), _outer_vertex(outer_vertex) {
+    _msg = "In ComposedGraph |_G| vertex |_outer_vertex| does not belongs to any U[i]";
```
Comment (tunyash) [inline src/composed-graph.cpp]: why not specify the id of the _outer_vertex in the message? The same applies to all the error messages in the exceptions.

File: src/composed-graph.h [hunk 2]
```diff
@@ -0,0 +1,129 @@
+#include "nice-path-decomposition.h"
+#include "path-decomposition.h"
+#include "centroid-tree.h"
+#include <string>
+#include <vector>
+
+#ifndef COMPOSED_GRAPH_H
```
Comment (tunyash) [inline src/composed-graph.h]: Best to move it to the very top of the file, before the includes

File: src/composed-graph.h [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    Graph _G;
+    std::vector<std::vector<vertex_t>> _U;
+    std::vector<NicePathDecomposition> _pw_u;
+    NicePathDecomposition _pw_compressed;
+
+};
+
+Graph get_good_subgraph(Graph &g, std::vector<vertex_t> &U_i);
+/*
+ * returning Graph is isomorphic to subgraph g[U_i] but have vertices from 0 to |U| - 1
```
Comment (tunyash) [inline src/composed-graph.h]: returning -> returned

File: src/composed-graph.h [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+ * This is used to create PathDecomposition from subgraph g[U]
+ * Function that rollbacks vertices indexes is PathDecomposition::enumerate()
```
Comment (tunyash) [inline src/composed-graph.h]: indexes -> indices

File: src/composed-test.cpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                {0, 1}
+        };
+
+        std::vector<std::vector<vertex_t>> U = {{0, 1}, {2, 3}};
+        std::vector<NicePathDecomposition> NPD(7);
+        ComposedGraph CG(g, U, NPD);
+        Graph res = CG.get_compressed_graph();
+        CHECK(get_set_edges(res) == need_edges);
+        CHECK(boost::num_vertices(res) == m);
+    } SUBCASE("Matching") {
```
Comment (tunyash) [inline src/composed-test.cpp]: Let's start each subcase in a new line.

File: src/composed-graph.cpp [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+NicePathDecomposition ComposedGraph::get_path_decomposition(NicePathDecomposition &pw_compressed) {
+    int n = boost::num_vertices(_G);
+    _pw_compressed = pw_compressed;
+
+    std::vector<bool> used(boost::num_vertices(_G), false);
+    std::vector<NicePathDecomposition::Bag> G_path_decomp;
+
+    for (auto bag : _pw_compressed.get_nice_bags()) {
+        if (bag.type != NicePathDecomposition::ADD_VERTEX) continue;
```
Comment (tunyash) [inline src/composed-graph.cpp:108]: What if we remove a node from the compressed path decomposition's bag? Shouldn't we remove some of the nodes in the corresponding block?

File: src/nice-path-decomposition.h [hunk 7]
```diff
@@ -41,22 +43,32 @@ class NicePathDecomposition : public PathDecomposition {
...
+        NiceBagsCorrectnessException(error_types type, edge_t edge);
+
+        NiceBagsCorrectnessException(error_types type);
 
         ~NiceBagsCorrectnessException() throw() = default;
 
         const char *what() const throw();
     };
 
+    void enumerate(std::vector<vertex_t> &U);
```
Comment (tunyash) [inline src/nice-path-decomposition.h:55]: What does this method do exactly?

File: src/composed-test.cpp [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+        std::vector<std::pair<int, int>> edges = {
+                {0, 1}, {1, 2}, {0, 2},
+                {3, 4}, {4, 5}, {3, 5},
+                {0, 3}, {1, 4}, {2, 5}
+        };
+        Graph g(edges.begin(), edges.end(), 6);
+        std::vector<std::vector<vertex_t>> U = {
+                {0, 1}, {3, 4}
+        };
+        std::vector<NicePathDecomposition> NPD(2);
```
Comment (tunyash) [inline src/composed-test.cpp:101]: What are the default path decompositions? You use them in all the tests, can you add some with nontrivial inner path decompositions?

**Review by tunyash — CHANGES_REQUESTED** [review]
My main concern is the uniformity of used inner path decompositions (they are all generated via default constructor and never actively filled out). Can you (a) add more tests with different inner path decompositions (b) explain what the default path decomposition is and how is it generated? Can't it be the reason you have to convert the indices of the nodes to 0...n?

File: src/composed-test.cpp (hunk 8 above)
Comment (elproblema) [reply src/composed-test.cpp:101]: This test is created to check correctness of exceptions. I don't need NicePathDecomposition for it

File: src/composed-graph.cpp (hunk 6 above)
Comment (elproblema) [reply src/composed-graph.cpp:108]: No, they are instantly removed after adding node from compressed graph

File: src/composed-test.cpp (hunk 8 above)
Comment (tunyash) [reply src/composed-test.cpp:101]: What about other tests? I did not spot any inner decompositions defined in another way.

File: src/composed-graph.cpp (hunk 6 above)
Comment (tunyash) [reply src/composed-graph.cpp:108]: Can you explain it further? I am not sure where it happens.

Comment (elproblema) [reply src/composed-graph.cpp:108]: I am just copying all bags from small NicePathDecompositions so all vertices that was declared at small NicePathDecomposition was removed there

File: src/composed-test.cpp (hunk 8 above)
Comment (elproblema) [reply src/composed-test.cpp:101]: Last subcase uses functions perform_path_decomposition from composed-graph.h. It uses CentroidTree

File: src/composed-graph.cpp (hunk 6 above)
Comment (tunyash) [reply src/composed-graph.cpp:108]: So when do you remove _matchings[in_bag.vertex] nodes then?

File: src/composed-test.cpp (hunk 8 above)
Comment (tunyash) [reply src/composed-test.cpp:101]: Ok, please add some more including the explicit ones (usually they are easier to follow for someone looking for usage examples)

**Review by tunyash — APPROVED** [review]
LGTM

File: src/composed-graph.cpp (hunk 6 above)
Comment (elproblema) [reply src/composed-graph.cpp:108]: When set with _matchings[in_bag.vertex] appears

**Pull request merged**